File: module/storage-kv/db_meta.go
package main

import (
	"github.com/pantopic/wazero-lmdb/sdk-go"
)

var (
	// Logical clock representing seconds of uptime since shard creation
	metaKeyEpoch = []byte(`epoch`)

	// Index of last applied raft log entry
	metaKeyIndex = []byte(`index`)

	// Autoincrement cursor for generating lease ids
	metaKeyLeaseID = []byte(`lease_id`)

	// Last applied data revision
	metaKeyRevision = []byte(`rev`)

	// Compaction cursor - Keys up to this revision have been compacted (always <= rev_min)
	metaKeyRevisionCompacted = []byte(`rev_compacted`)

	// Compaction target - Keys up to this revision are no longer visible
	metaKeyRevisionMin = []byte(`rev_min`)

	// Shard raft term - Prevents duplicate controllers
	metaKeyTerm = []byte(`term`)
)

type dbMetaImpl struct {
	db
}

func (db dbMetaImpl) init(txn *lmdb.Txn) (index uint64) {
	var err error
	db.open(txn)
	for _, k := range [][]byte{
		metaKeyEpoch,
		metaKeyIndex,
		metaKeyLeaseID,
		metaKeyRevisionCompacted,
		metaKeyTerm,
	} {
		if _, err = db.db.getUint64(txn, k); lmdb.IsNotFound(err) {
			err = db.putUint64(txn, k, 0)
		}
		if err != nil {
			panic(err)
		}
	}
	for _, k := range [][]byte{
		metaKeyRevision,
		metaKeyRevisionMin,
	} {
		if _, err = db.db.getUint64(txn, k); lmdb.IsNotFound(err) {
			err = db.putUint64(txn, k, 1)
		}
		if err != nil {
			return
		}
	}
	if index, err = db.getIndex(txn); err != nil {
		panic(err)
	}
	return
}

func (db dbMetaImpl) getEpoch(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyEpoch)
}

func (db dbMetaImpl) setEpoch(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyEpoch, val)
}

func (db dbMetaImpl) getIndex(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyIndex)
}

func (db dbMetaImpl) setIndex(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyIndex, val)
}

func (db dbMetaImpl) getLeaseID(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyLeaseID)
}

func (db dbMetaImpl) setLeaseID(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyLeaseID, val)
}

func (db dbMetaImpl) getRevision(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyRevision)
}

func (db dbMetaImpl) setRevision(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyRevision, val)
}

func (db dbMetaImpl) getRevisionCompacted(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyRevisionCompacted)
}

func (db dbMetaImpl) setRevisionCompacted(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyRevisionCompacted, val)
}

func (db dbMetaImpl) getRevisionMin(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyRevisionMin)
}

func (db dbMetaImpl) setRevisionMin(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyRevisionMin, val)
}

func (db dbMetaImpl) getTerm(txn *lmdb.Txn) (val uint64, err error) {
	return db.getUint64(txn, metaKeyTerm)
}

func (db dbMetaImpl) setTerm(txn *lmdb.Txn, val uint64) (err error) {
	return db.putUint64(txn, metaKeyTerm, val)
}

File: module-zig/storage-kv/db_meta.zig
//! Mirrors module/storage-kv/db_meta.go

const std = @import("std");
const mdb = @import("mdb");

const Db = @import("db.zig").Db;

/// Logical clock representing seconds of uptime since shard creation
const meta_key_epoch = "epoch";

/// Index of last applied raft log entry
const meta_key_index = "index";

/// Autoincrement cursor for generating lease ids
const meta_key_lease_id = "lease_id";

/// Last applied data revision
const meta_key_revision = "rev";

/// Compaction cursor - Keys up to this revision have been compacted (always <= rev_min)
const meta_key_revision_compacted = "rev_compacted";

/// Compaction target - Keys up to this revision are no longer visible
const meta_key_revision_min = "rev_min";

/// Shard raft term - Prevents duplicate controllers
const meta_key_term = "term";

pub const DbMeta = struct {
    db: Db,

    pub fn init(self: DbMeta, txn: mdb.Txn) u64 {
        self.db.open(txn);
        for ([_][]const u8{
            meta_key_epoch,
            meta_key_index,
            meta_key_lease_id,
            meta_key_revision_compacted,
            meta_key_term,
        }) |k| {
            _ = self.db.getUint64(txn, k) catch |err| {
                if (err != mdb.Error.NotFound) {
                    std.debug.panic("{s}", .{@errorName(err)});
                }
                self.db.putUint64(txn, k, 0) catch |perr|
                    std.debug.panic("{s}", .{@errorName(perr)});
            };
        }
        for ([_][]const u8{
            meta_key_revision,
            meta_key_revision_min,
        }) |k| {
            _ = self.db.getUint64(txn, k) catch |err| {
                if (err != mdb.Error.NotFound) {
                    return 0;
                }
                self.db.putUint64(txn, k, 1) catch {
                    return 0;
                };
            };
        }
        return self.getIndex(txn) catch |err|
            std.debug.panic("{s}", .{@errorName(err)});
    }

    pub fn getEpoch(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_epoch);
    }

    pub fn setEpoch(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_epoch, val);
    }

    pub fn getIndex(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_index);
    }

    pub fn setIndex(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_index, val);
    }

    pub fn getLeaseID(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_lease_id);
    }

    pub fn setLeaseID(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_lease_id, val);
    }

    pub fn getRevision(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_revision);
    }

    pub fn setRevision(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_revision, val);
    }

    pub fn getRevisionCompacted(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_revision_compacted);
    }

    pub fn setRevisionCompacted(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_revision_compacted, val);
    }

    pub fn getRevisionMin(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_revision_min);
    }

    pub fn setRevisionMin(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_revision_min, val);
    }

    pub fn getTerm(self: DbMeta, txn: mdb.Txn) !u64 {
        return self.db.getUint64(txn, meta_key_term);
    }

    pub fn setTerm(self: DbMeta, txn: mdb.Txn, val: u64) !void {
        return self.db.putUint64(txn, meta_key_term, val);
    }
};
